Web Inspector: Show grid/flex overlays when in element selection

dcrousso opened the pull request:
#### f51cad26306748dfdef688967af230eed8484136
<pre>
Web Inspector: Show grid/flex overlays when in element selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=251737">https://bugs.webkit.org/show_bug.cgi?id=251737</a>
&lt;rdar://problem/105041619&gt;

Reviewed by Patrick Angle.

This allows for developers to more easily visually identify the organization of a node&apos;s contents using element select mode (e.g. hovering over a `gap` area will now be more recognizable due to the hatching shown by the flex overlay).

* Source/JavaScriptCore/inspector/protocol/DOM.json:
Move all the configuration options for grid and flex overlays to bespoke protocol objects.

* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
Ensure that these protocol objects have helper constants for pulling values out of them.

* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::paint):
(WebCore::InspectorOverlay::getHighlight):
(WebCore::InspectorOverlay::shouldShowOverlay const):
(WebCore::InspectorOverlay::showHighlightGridOverlayForNode): Added.
(WebCore::InspectorOverlay::hideHighlightGridOverlay): Added.
(WebCore::InspectorOverlay::showHighlightFlexOverlayForNode): Added.
(WebCore::InspectorOverlay::hideHighlightFlexOverlay): Added.
In addition to the always-on grid/flex overlay data, have another member variable for a single grid/flex overlay that takes precedence (and is drawn on top) of all other grid/flex overlays.

* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::parseRequiredConfigColor): Renamed from `parseConfigColor`.
(WebCore::parseOptionalConfigColor): Added.
(WebCore::InspectorDOMAgent::willDestroyFrontendAndBackend):
(WebCore::InspectorDOMAgent::handleTouchEvent):
(WebCore::InspectorDOMAgent::inspect):
(WebCore::InspectorDOMAgent::highlightMousedOverNode):
(WebCore::InspectorDOMAgent::setSearchingForNode):
(WebCore::InspectorDOMAgent::highlightConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject): Added.
(WebCore::InspectorDOMAgent::flexOverlayConfigFromInspectorObject): Added.
(WebCore::InspectorDOMAgent::setInspectModeEnabled):
(WebCore::InspectorDOMAgent::showGridOverlay):
(WebCore::InspectorDOMAgent::showFlexOverlay):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.get defaultLayoutOverlayColor): Added.
(WI.DOMNode.prototype.showLayoutOverlay):
Handle changes to `DOM.showGridOverlay` and `DOM.showFlexOverlay`.

* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.set inspectModeEnabled):
Pass along the current grid/flex overlay configuration when enabling element selection mode.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):
Add some on-by-default `WI.Setting` to control this behavior.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/dom/showFlexOverlay.html:
* LayoutTests/inspector/dom/showGridOverlay.html:

Canonical link: <a href="https://commits.webkit.org/259989@main">https://commits.webkit.org/259989@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fbbbf1c530c945d756d7013c18ea2cef03007792

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115677 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115336 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110400 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6735 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98686 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12918 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95889 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40474 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82191 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96069 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8741 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28885 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95508 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6636 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9279 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5458 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-pseudo/first-letter-with-span.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30642 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14897 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48431 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104250 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6906 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10822 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25831 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -432,15 +432,31 @@ void InspectorOverlay::paint(GraphicsContext& context)
     }
 
     for (const InspectorOverlay::Grid& gridOverlay : m_activeGridOverlays) {
+        if (m_specialGridOverlay && gridOverlay.gridNode == m_specialGridOverlay->gridNode)
```
Instead of `m_specialXxxOverlay` can we use a name like `m_highlightedElementGridOverlay` or something else a bit more descriptive to make it easier to understand _why_ it is special?

patrickangle commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -607,6 +650,23 @@ bool InspectorOverlay::removeGridOverlayForNode(Node& node)
     });
 }
 
+void InspectorOverlay::showSpecialGridOverlayForNode(Node& node, const InspectorOverlay::Grid::Config& gridOverlayConfig)
```
Same question here about naming using `special` instead of something more descriptive?

patrickangle commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -329,10 +334,13 @@ void InspectorDOMAgent::willDestroyFrontendAndBackend(Inspector::DisconnectReaso
 
     Protocol::ErrorString ignored;
-    setSearchingForNode(ignored, false, nullptr, false);
+    setSearchingForNode(ignored, false, nullptr, nullptr, nullptr, false);
```
Aside: this is an unfortunately opaque set of parameters 😅

dcrousso replied:
luckily it's in the same file, so looking at the definition is pretty easy

but yeah would be more ideal to have some sort of clarity in there (especially in a way that conforms to WebKit style)

patrickangle commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -1293,31 +1344,63 @@ std::unique_ptr<InspectorOverlay::Highlight::Config> InspectorDOMAgent::highligh
     Protocol::ErrorString errorString;
 
-    setSearchingForNode(errorString, enabled, WTFMove(highlightConfig), showRulers && *showRulers);
+    setSearchingForNode(errorString, enabled, WTFMove(highlightConfig), WTFMove(gridOverlayConfig), WTFMove(flexOverlayConfig), showRulers && *showRulers);
```
Aside: This patch really makes me dislike names `InspectMode` and `SearchingForNode` since neither of those are what we call the features (`ElementSelection`) 😞

dcrousso replied:
yeah we have a few naming discrepancies like this.  definitely would be worth a followup to fix all the naming :D

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
```diff
@@ -681,11 +681,29 @@ WI.DOMManager = class DOMManager extends WI.Object
         };
+        if (WI.settings.showGridOverlayDuringElementSelection.value) {
+            commandArguments.gridOverlayConfig = {
+                gridColor: WI.DOMNode.defaultLayoutOverlayColor.toProtocol(),
```
I wonder if it will be confusing that we use one of the five default colors, effectively meaning there is a 4/5 chance that the grid/flex color won't match the one in the Layout sidebar. That said, we can't really resolve the color for any given node here because we don't know what node someone will hover. Have you considered perhaps using grey color to give a sense of the overlay without always using magenta? Maybe some screenshots would convince me this is fine? (I'm lacking spare cycles to build this patch myself right now 😬)

dcrousso replied:
that's a fair point.  i figured that if the magenta was good enough to be our initial option, it'd be good enough for any node 😅

i tried a few other colors like grey, but they all felt a bit too muted and not really very "stand out", whereas the magenta is a super nice contrast with the blue color we use for content (and green for padding)

while i do agree that maybe using the same color that could match with an already enabled/showing overlay could be a bit confusing for the user, if it's ok with you i'd rather handle that problem once this functionality is landed as changing the color in the frontend is _way_ easier than modifying the entire backend :)

patrickangle replied:
After looking at this, I think you are right that magenta is a good color here given the other colors in use. Thank you for explaining your reasoning here!

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -648,12 +653,19 @@ WI.DOMNode = class DOMNode extends WI.Object
+                showAreaNames: WI.settings.gridOverlayShowAreaNames.value,
+            };
+
+            // COMPATIBILITY (macOS 13.0, iOS 16.0): DOM.GridOverlayConfig did not exist yet.
```
These will be `macOS 13.?, iOS 16.?` right now...

patrickangle replied:
Also from a correctness point of view should we conditionalize mapping the arguments out to top-level keys to only occur when `DOM.GridOverlayConfig` doesn't exist? That way we don't accidentally start relying on this behavior somehow.

dcrousso replied:
FYI we unfortunately only generate JS for enums, not any other types :/

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -666,8 +678,15 @@ WI.DOMNode = class DOMNode extends WI.Object
+                showOrderNumbers: WI.settings.flexOverlayShowOrderNumbers.value,
+            };
+
+            // COMPATIBILITY (macOS 13.0, iOS 16.0): DOM.FlexOverlayConfig did not exist yet.
```
Ditto both my comments on :665

patrickangle reviewed (commented):
For the most part this looks great - just not convinced that using the first default color for every element is the best choice yet, but happy to be proven wrong.

dcrousso committed:
Web Inspector: Show grid/flex overlays when in element selection
https://bugs.webkit.org/show_bug.cgi?id=251737
<rdar://problem/105041619>

Reviewed by Patrick Angle.

This allows for developers to more easily visually identify the organization of a node's contents using element select mode (e.g. hovering over a `gap` area will now be more recognizable due to the hatching shown by the flex overlay).

* Source/JavaScriptCore/inspector/protocol/DOM.json:
Move all the configuration options for grid and flex overlays to bespoke protocol objects.

* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
Ensure that these protocol objects have helper constants for pulling values out of them.

* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::paint):
(WebCore::InspectorOverlay::getHighlight):
(WebCore::InspectorOverlay::shouldShowOverlay const):
(WebCore::InspectorOverlay::showHighlightGridOverlayForNode): Added.
(WebCore::InspectorOverlay::hideHighlightGridOverlay): Added.
(WebCore::InspectorOverlay::showHighlightFlexOverlayForNode): Added.
(WebCore::InspectorOverlay::hideHighlightFlexOverlay): Added.
In addition to the always-on grid/flex overlay data, have another member variable for a single grid/flex overlay that takes precedence (and is drawn on top) of all other grid/flex overlays.

* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::parseRequiredConfigColor): Renamed from `parseConfigColor`.
(WebCore::parseOptionalConfigColor): Added.
(WebCore::InspectorDOMAgent::willDestroyFrontendAndBackend):
(WebCore::InspectorDOMAgent::handleTouchEvent):
(WebCore::InspectorDOMAgent::inspect):
(WebCore::InspectorDOMAgent::highlightMousedOverNode):
(WebCore::InspectorDOMAgent::setSearchingForNode):
(WebCore::InspectorDOMAgent::highlightConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject): Added.
(WebCore::InspectorDOMAgent::flexOverlayConfigFromInspectorObject): Added.
(WebCore::InspectorDOMAgent::setInspectModeEnabled):
(WebCore::InspectorDOMAgent::showGridOverlay):
(WebCore::InspectorDOMAgent::showFlexOverlay):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.get defaultLayoutOverlayColor): Added.
(WI.DOMNode.prototype.showLayoutOverlay):
Handle changes to `DOM.showGridOverlay` and `DOM.showFlexOverlay`.

* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.set inspectModeEnabled):
Pass along the current grid/flex overlay configuration when enabling element selection mode.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):
Add some on-by-default `WI.Setting` to control this behavior.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/dom/showFlexOverlay.html:
* LayoutTests/inspector/dom/showGridOverlay.html:

Canonical link: https://commits.webkit.org/259989@main
